speculos/mcu/display.py

```diff
@@ -99,34 +71,67 @@ class FrameBuffer:
         self.recreate_public_screenshot = True
         self.model = model
         self.current_screen_size = MODELS[model].screen_size
-        self.screenshot = Screenshot(self.current_screen_size)
-        # Init published content now, don't wait for the first request
-        if self.model == "stax":
-            self.update_public_screenshot()
+        self._width, self._height = MODELS[model].screen_size
 
-    def draw_point(self, x: int, y: int, color: int) -> None:
+    @cache
+    def check_color(self, color: int) -> int:
         # There are only 2 colors on the Nano S and the Nano X but the one
         # passed in argument isn't always valid. Fix it here.
         if self.model != 'stax':
             if color != 0x000000:
                 color = FrameBuffer.COLORS.get(self.model, color)
-        self.pixels[(x, y)] = color
+        return color
+
+    def draw_point(self, x: int, y: int, color: int) -> None:
+        self.pixels[(x, y)] = self.check_color(color)
+
+    def draw_horizontal_line(self, x0: int, y: int, width: int, color: int) -> None:
+        for x in range(x0, x0 + width):
+            self.pixels[(x, y)] = self.check_color(color)
+
+    def draw_rect(self, x0: int, y0: int, width: int, height: int, color: int) -> None:
+        color = self.check_color(color)
+
+        if x0 == 0 and y0 == 0 and width == self._width and height == self._height:
+            self.default_color = color
+            self.draw_default_color = True
+            self.pixels = {}
+            self.screenshot_pixels = {}
+            return
+
+        for x in range(x0, x0 + width):
+            for y in range(y0, y0 + height):
+                self.pixels[(x, y)] = color
+
+    def _get_image(self) -> bytes:
+        data = bytearray(self.default_color.to_bytes(3, "big")) * self._width * self._height
+        for (x, y), color in self.screenshot_pixels.items():
+            pos = 3 * (y * self._width + x)
+            data[pos:pos + 3] = color.to_bytes(3, "big")
+        return bytes(data)
+
+    def _get_screenshot_iobytes_value(self) -> bytes:
+        # Get the pixels object once, as it may be replaced during the loop.
+        data = self._get_image()
 
-    def screenshot_update_pixels(self):
-        # Update the screenshot object with our current pixels content
-        self.screenshot.update(self.pixels)
+        image = Image.frombytes("RGB", self.current_screen_size, data)
+        iobytes = io.BytesIO()
+        image.save(iobytes, format="PNG")
+        return iobytes.getvalue()
 
     def take_screenshot(self) -> Tuple[Tuple[int, int], bytes]:
-        self.current_screen_size, self.current_data = self.screenshot.get_image()
-        return self.current_screen_size, self.current_data
+        return self.current_screen_size, self._get_image()
 
     def update_screenshot(self) -> None:
-        self.screenshot.update(self.pixels)
+        self.screenshot_pixels = {**self.screenshot_pixels, **self.pixels}
```

Comment:
```suggestion
        self.screenshot_pixels.update(self.pixels)
```

Reply:
Not sure, but in previous version there was a comment saying:
`# Don't call update, replace the object instead`
I trusted it, without testing the performance...